karabo_data/reader.py

```diff
@@ -1324,18 +1313,22 @@ def stack_detector_data(train, data, axis=-3, modules=16, only='', xcept=(),
         s1, s2, *_ = sorted(shapes)
         if len(shapes) > 2 or (s1[0] != 0) or (s1[1:] != s2[1:]):
             raise ValueError("Arrays have mismatched shapes: {}".format(shapes))
-        skip = {n for n, a in modno_arrays.items() if a.shape == s1}
+        empty_mods = {n for n, a in modno_arrays.items() if a.shape == s1}
+        for modno in empty_mods:
+            del modno_arrays[modno]
         shapes.remove(s1)
     if max(modno_arrays) >= modules:
         raise IndexError("Module {} is out of range for a detector with {} modules"
                          .format(max(modno_arrays), modules))
 
     dtype = dtypes.pop()
     shape = shapes.pop()
-    combined = np.full((modules,) + shape, fillvalue, dtype=dtype)
-    for modno, array in modno_arrays.items():
-        if modno in skip:
-            continue
-        combined[modno] = array
-
-    return np.moveaxis(combined, 0, axis)
+    if virtual:
+        from .read_machinery import VirtualStack
+        return VirtualStack(modno_arrays, modules, shape, dtype, fillvalue)
+    else:
+        combined = np.full((modules,) + shape, fillvalue, dtype=dtype)
+        for modno, array in modno_arrays.items():
+            combined[modno] = array
+
+        return np.moveaxis(combined, 0, axis)
```

Comment:
I would use something like 

```py
v = VirtualStack(...)`
if virtual:
    return v
return v.asarray()
```
to avoid the duplication of code.

Reply:
Yup, that makes sense.

Reply:
Actually, I remember now why I didn't do this: they're logically equivalent, but different in terms of memory arrangement. The existing `stack_detector_data` code keeps the data from each module together in memory but then adjusts the axes to look like (pulses, modules, slow_scan, fast_scan). `VirtualStack.asarray()` reshuffles the data so each pulse is contiguous in memory.

I'm not sure how big a difference this makes, but I'll see if it's practical to make `VirtualStack.asarray()` do something more similar to the existing assembly.

Reply:
I think the difference depends on the way of looping over the pixel values. Normally, one pick a pulse (in a for loop) and then loop over the pixels, then it would be nice to have all the pixels continuous in memory. Otherwise, it will need to jump to the memory of another module 16 times for one pulse and I guess the performance penalty is 10% - 20%.

Aha, here what you do is copy to the stacked data to the assembled image. Then which one is faster depends on your assembling code,